build.assets/tooling/lib/github/workflows.go

```diff
@@ -0,0 +1,151 @@
+		}
+	}
+	return true
+}
+
+// Insert adds an element to a set. Adding an element multiple times is not
+// considered an error.
+func (s RunIDSet) Insert(runID int64) {
+	s[runID] = struct{}{}
+}
+
+// NotIn returns the elements that are in `s` but not in `other`
+func (s RunIDSet) NotIn(other RunIDSet) RunIDSet {
+	result := make(RunIDSet)
+	for k := range s {
+		if _, present := other[k]; !present {
+			result[k] = struct{}{}
+		}
+	}
+	return result
+}
+
+// WorkflowRuns defines the minimal API used to lst and query GitHub action
+// runner workflows and jobs
+type WorkflowRuns interface {
+	GetWorkflowRunByID(ctx context.Context, owner, repo string, runID int64) (*github.WorkflowRun, *github.Response, error)
+	ListWorkflowRunsByFileName(ctx context.Context, owner, repo, workflowFileName string, opts *github.ListWorkflowRunsOptions) (*github.WorkflowRuns, *github.Response, error)
+}
+
+// ListWorkflowRuns returns a set of RunIDs, representing the set of all for
+// workflow runs created since the supplied start time.
+func ListWorkflowRuns(ctx context.Context, actions WorkflowRuns, owner, repo, path, ref string, since time.Time) (RunIDSet, error) {
+	listOptions := github.ListWorkflowRunsOptions{
+		ListOptions: github.ListOptions{
+			PerPage: 100,
+		},
+		Branch:  ref,
+		Created: ">" + since.Format(time.RFC3339),
+	}
+
+	runIDs := make(RunIDSet)
+
+	for {
+		runs, resp, err := actions.ListWorkflowRunsByFileName(ctx, owner, repo, path, &listOptions)
+		if err != nil {
+			return nil, trace.Wrap(err, "Failed to fetch runs")
+		}
+
+		for _, r := range runs.WorkflowRuns {
+			runIDs.Insert(r.GetID())
+		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+
+		listOptions.Page = resp.NextPage
+	}
+
+	return runIDs, nil
+}
+
+// WorkflowJobLister defines the minimal GitHub client interafce required to
+// list query and compose workflow jobs.
+type WorkflowJobLister interface {
+	ListWorkflowJobs(ctx context.Context, owner, repo string, runID int64, opts *github.ListWorkflowJobsOptions) (*github.Jobs, *github.Response, error)
+}
+
+// ListWorkflowJobs lists the jobs for a given workflow run in the specified
+// repository.
+func ListWorkflowJobs(ctx context.Context, lister WorkflowJobLister, owner, repo string, runID int64) ([]*github.WorkflowJob, error) {
+	listOptions := github.ListWorkflowJobsOptions{}
+	result := []*github.WorkflowJob{}
+	for {
+		jobs, resp, err := lister.ListWorkflowJobs(ctx, owner, repo, runID, &listOptions)
+		if err != nil {
+			return nil, trace.Wrap(err, "Failed to fetch workflow jobs")
+		}
+
+		result = append(result, jobs.Jobs...)
+
+		if resp.NextPage == 0 {
+			break
+		}
+
+		listOptions.Page = resp.NextPage
+	}
+
+	return result, nil
+}
+
+// WaitForRun blocks until the specified workflow run completes, and returns the overall
+// workflow status.
+func WaitForRun(ctx context.Context, actions WorkflowRuns, owner, repo, path, ref string, runID int64) (string, error) {
+	ticker := time.NewTicker(30 * time.Second)
```

Comment:
That's pretty big lag, maybe we can decrease it? (and don't print status on every check)

Reply:
I'm happy to reduce it, but remember that we have a build times in the order of tens of minutes - so an average 15s increase is not going to hurt much.

I'm torn on the status one. I think we need _some_ sort of pacifier. If I saw a tool emitting no output for 10 minutes I'd think it crashed. Perhaps output once a minute or something?